### PR TITLE
ieee802154/hal: adapt frame filter and source address matching changes

### PR DESCRIPTION
**Pull request by jia200x**
### Contribution description
This PR implements the proposal from https://github.com/RIOT-OS/RIOT/pull/13943#issuecomment-724699696 (in a slightly different way, as described above).

This removes the `set_rx_mode` and `set_hw_addr_filter` and replace them with 3 functions:
- `config_addr_filter`
- `set_frame_filter_mode`
- `config_src_addr_match`

The original proposal didn't include the `set_frame_filter_mode`, but it turns out to be much better IMO to split the original proposal of `config_addr_filter` because there's indeed a mode for the frame filter (promisc, sniffer, accept only ACK frames, etc). This also adds a "reject all frames" mode, that might be useful for certain stacks (e.g OpenWSN).

Also, a `CONFIG_IEEE802154_AUTO_ACK_DISABLE macro was added in order to disable Auto ACK in radios (required for OpenWSN)

### Testing procedure
- Test that `examples/gnrc_networking` runs properly with `netdev_ieee802154_submac`.
- OpenWSN should also be tested to ensure that everything works ok

### Issues/PRs references
#13943

### CODE REVIEW

**Comment by fjmolinas** [comment]
> This also adds a "reject all frames" mode, that might be useful for certain stacks (e.g OpenWSN).

Where did you have in mind?

**Comment by fjmolinas** [comment]
> - `config_addr_filter`
> - `set_frame_filter_mode`
> - `config_src_addr_match`

+1 for splitting.

**Comment by fjmolinas** [comment]
Why change to a compile flag for `CONFIG_IEEE802154_AUTO_ACK_DISABLE`?

**Comment by jia200x** [comment]
> Where did you have in mind?

Basically this disabled SFD detection. So, it can be used to disable RX without having to turn off the transceiver. An upper layer can benefit in cases where fast transitions need to be done.

I also think it's useful for helping the MAC layer to be in a defined state (e.g prevent receiving frames when moving from RX_ON to TX_ON)

**Comment by jia200x** [comment]
> Why change to a compile flag for CONFIG_IEEE802154_AUTO_ACK_DISABLE?

The reason for this change is because the MAC layer has control of whether a confirmable frame should be sent or not. This feature only disables the AutoACK feature, required by MAC layers that implement stuff out of the scope of the IEEE 802.15.4 2006 standard (e.g IEEE 802.15.4e). Otherwise, an upper layer would not set the ACK_REQ bit when sending frames.

**Comment by fjmolinas** [comment]
> Basically this disabled SFD detection. So, it can be used to disable RX without having to turn off the transceiver. An upper layer can benefit in cases where fast transitions need to be done.
> 
> I also think it's useful for helping the MAC layer to be in a defined state (e.g prevent receiving frames when moving from RX_ON to TX_ON)

Is it the same as disabling SFD? Why not have a radio state where RX is on but not listening? It seems that "reject all frames" is a more convoluted, more ambiguous way of saying the same thing no?

**Comment by jia200x** [comment]
> Is it the same as disabling SFD? Why not have a radio state where RX is on but not listening? It seems that "reject all frames" is a more convoluted, more ambiguous way of saying the same thing no?

I think this is the idea. How should we name it in order to reflect that?

**Comment by benpicco** [comment]
> Why not have a radio state where RX is on but not listening? 

Why would RX be on if we are not listening?
How does this state differ from `TRX OFF`?

**Comment by jia200x** [comment]
> Why would RX be on if we are not listening?
How does this state differ from TRX OFF?

The TRX_OFF state turns off the PLL. There's an associated turnaround time.
The "disable SFD/reject frames" procedure simply disables passing stuff to the framebuffer and indicating a frame.
This can be used for CCA or for moments when the transceiver must be disabled but the turnaround time might be an issue (e.g a MAC layer can disable SFD detection right before switching to TRX_OFF, without worrying about received frames)

**Comment by benpicco** [comment]
This sounds like what AT86RF215 calls TXPREP

at86rf215

**Comment by jia200x** [comment]
> or the PLL_ON state on AT86RF233

I think it's not intended to be a "transceiver state" (e.g PLL_ON is mapped to TX_ON if the radio is running in Basic Mode). I think for this radio it would map to "RX_PDT_DIS (register 0x15). This simply disables SFD detection, while the radio is still able to e.g do CCA (which requires RX_ON).

In any case I introduced it as an experimental feature that could be exploited by some MAC layers. But it's absolutely not mandatory to have that one and we could go without it

**Comment by jia200x** [comment]
I removed the FRAME_REJECT option and rebased to current master
This PR is important in order to simplify the implementation of some radio HALs and to enable features required by the IEEE 802.15.4 MAC layer.
Therefore, would be nice if this can be reviewed soon :)

**Comment by fjmolinas** [comment]
> > or the PLL_ON state on AT86RF233
> 
> I think it's not intended to be a "transceiver state" (e.g PLL_ON is mapped to TX_ON if the radio is running in Basic Mode). I think for this radio it would map to "RX_PDT_DIS (register 0x15). This simply disables SFD detection, while the radio is still able to e.g do CCA (which requires RX_ON).
> 
> In any case I introduced it as an experimental feature that could be exploited by some MAC layers. But it's absolutely not mandatory to have that one and we could go without it

I'm fine with using the `reject-all` for now to move forward, what do you think @benpicco

**Comment by fjmolinas** [comment]
> > > or the PLL_ON state on AT86RF233
> > 
> > 
> > I think it's not intended to be a "transceiver state" (e.g PLL_ON is mapped to TX_ON if the radio is running in Basic Mode). I think for this radio it would map to "RX_PDT_DIS (register 0x15). This simply disables SFD detection, while the radio is still able to e.g do CCA (which requires RX_ON).
> > In any case I introduced it as an experimental feature that could be exploited by some MAC layers. But it's absolutely not mandatory to have that one and we could go without it
> 
> I'm fine with using the `reject-all` for now to move forward, what do you think @benpicco

@maribu and @benpicco might want to take a look, otherwise, I can test tomorrow.

**Comment by jia200x** [comment]
> I'm fine with using the reject-all for now to move forward, what do you think @benpicco

I removed it in the end. In case we need it, we can add it later :)

**Comment by fjmolinas** [comment]
>- Test that `examples/gnrc_networking` runs properly with `netdev_ieee802154_submac`.

Tested on nrf52840-mdk

```
2021-06-15 14:17:43,191 # --- fe80::a804:8169:befa:4b38 PING statistics ---
2021-06-15 14:17:43,197 # 100 packets transmitted, 100 packets received, 0% packet loss
2021-06-15 14:17:43,201 # round-trip min/avg/max = 5.215/8.042/23.319 ms
```

**Comment by fjmolinas** [comment]
Tested `OpenWSN` on `openmote-b`:

```
15:09:50 VERBOSE 7d12 [FRAG] fragmenting a big packet, original size 87, number of fragments 2
15:09:51 SUCCESS [GATEWAY] Reassembled 2 frags with tag 145 into an IPv6 packet of size 87
15:09:52 SUCCESS [GATEWAY] Reassembled 2 frags with tag 134 into an IPv6 packet of size 87
15:09:53 SUCCESS [GATEWAY] Reassembled 2 frags with tag 135 into an IPv6 packet of size 87
15:09:54 SUCCESS [GATEWAY] Reassembled 2 frags with tag 136 into an IPv6 packet of size 87
15:09:55 SUCCESS [GATEWAY] Reassembled 2 frags with tag 137 into an IPv6 packet of size 87
15:09:56 SUCCESS [GATEWAY] Reassembled 2 frags with tag 138 into an IPv6 packet of size 87
15:09:56 INFO received RPL DAO from bbbb:0:0:0:ae8d:fee1:208c:870d
        - parents:
           bbbb:0:0:0:ae8d:fee1:60c7:cb8a
        - children:
15:09:57 VERBOSE 7d12 [FRAG] fragmenting a big packet, original size 89, number of fragments 2
15:09:58 SUCCESS [GATEWAY] Reassembled 2 frags with tag 146 into an IPv6 packet of size 89
15:09:58 INFO received RPL DAO from bbbb:0:0:0:ae8d:fee1:e052:d7d7
        - parents:
           bbbb:0:0:0:ae8d:fee1:60c7:cb8a
        - children:
           bbbb:0:0:0:ae8d:fee1:e02c:e22d
15:09:59 SUCCESS [GATEWAY] Reassembled 2 frags with tag 140 into an IPv6 packet of size 87
15:10:03 INFO received RPL DAO from bbbb:0:0:0:ae8d:fee1:208c:870d
        - parents:
           bbbb:0:0:0:ae8d:fee1:60c7:cb8a
        - children:
15:11:29 INFO received RPL DAO from bbbb:0:0:0:ae8d:fee1:e02c:e22d                                                                                                                                                                                                                        - parents:                                                                                                             
           bbbb:0:0:0:ae8d:fee1:60c7:cb8a
        - children:
15:11:30 INFO received RPL DAO from bbbb:0:0:0:ae8d:fee1:208c:870d
        - parents:
           bbbb:0:0:0:ae8d:fee1:60c7:cb8a
        - children:
```

I seem to have issues with nrf52840 though, but they seem to be in master as well...

**Comment by jia200x** [comment]
> Note that this was tested without channel hopping since there is a separate issue there with nrf52840.

I will give it a look with the SDR to check what's going on. I also found once some weird issues with this radio (spurious transmissions).

Luckily https://github.com/RIOT-OS/RIOT/pull/15761/files (almost there!) will help us to detect these issues.

**Review by fjmolinas — APPROVED** [review]
ACK, issues experienced with `nrf52` and OpenWSN seem unrelated to this PR

**Comment by jia200x** [comment]
rebased!

**Comment by jia200x** [comment]
oh, it conflicts now with OpenWSN. I will need to rebase again :(

**Comment by jia200x** [comment]
done!

File: cpu/cc2538/radio/cc2538_rf.c [hunk 1]
```diff
@@ -174,6 +174,11 @@ void cc2538_init(void)
     /* setup mac timer */
     _cc2538_setup_mac_timer();
 
+    /* Enable Auto ACK */
+    if (IS_USED(MODULE_IEEE802154_RADIO_HAL)) {
```
Comment (jia200x) [inline cpu/cc2538/radio/cc2538_rf.c:178]: I just noticed this module doesn't exists anymore... I will need to adapt it

Comment (jia200x) [reply cpu/cc2538/radio/cc2538_rf.c]: done. I simply removed it. The introduced configuration file can be used for both netdev and the Radio HAL

**Comment by fjmolinas** [comment]
Please squash!

**Pull request merged**